nchelpers/__init__.py

```diff
@@ -0,0 +1,400 @@
+    #         if not hasattr(nc, key):
+    #             missing.append(key)
+    #     if missing:
+    #         raise ValueError(required_nc_attributes_msg.format(required, nc.filepath(), missing))
+    #
+    #     return {
+    #         to_: getattr(nc, from_)
+    #         for from_, to_ in map_.items()
+    #         }
+    #
+    # def file_metadata(nc):
+    #     """Return important global attributes from this file"""
+    #     if self.project_id == 'CMIP5':
+    #         meta = _get_file_metadata(nc, global_to_res_map_cmip5)
+    #     else:
+    #         meta = _get_file_metadata(nc, global_to_res_map_cmip3)
+    #
+    #     # Which variable(s) does this file contain?
+    #     meta['var'] = '+'.join(get_important_varnames(nc))  # just do the computation where needed
+    #
+    #     # Compute time metadata from the time value
+    #     time = get_timeseries(nc)
+    #     meta['tres'] = get_time_resolution(time['numeric'], time['units'])  # == self.time_resolution
+    #     tmin, tmax = get_time_range(nc)
+    #     tmin, tmax = num2date([tmin, tmax], time['units'], time['calendar'])
+    #     meta['trange'] = format_time_range(tmin, tmax, meta['tres'])  # == self.time_range_formatted
+    #
+    #     return meta
+
+    class UnifiedMetadata(object):
+        """Presents a unified interface to certain global metadata attributes in a CFDataset object.
+        Why?
+        - A CFDataset can have metadata attributes named according to CMIP3 or CMIP5 standards, depending on the file's
+          origin (which is indicated by project_id).
+        - We want a common interface, i.e., common names, for a selected set of those differently named attributes.
+        - We must avoid shadowing existing properties and methods on a CFDataset (or really, a netCDF4.Dataset) object
+          with the unified names we'd like to use for these metadata properties.
+        - We'd like to present them as properties instead of as a dict, which has uglier syntax
+        How?
+        - Create a property called metadata on CFDataset that is an instance of this class.
+        """
+        # TODO: Make this a singleton
+
+        def __init__(self, dataset):
+            self.dataset = dataset
+
+        _aliases = {
+            'institution': {
+                'CMIP3': 'institute',
+                'CMIP5': 'institute_id',
+            },
+            'model': {
+                'CMIP3': 'source',
+                'CMIP5': 'model_id',
+            },
+            'emissions': {
+                'CMIP3': 'experiment_id',
+                'CMIP5': 'experiment_id',
+            },
+            'run': {
+                'CMIP3': 'realization',
+                'CMIP5': 'parent_experiment_rip',
+            },
+            'project': {
+                'CMIP3': 'project_id',
+                'CMIP5': 'project_id',
+            },
+        }
+
+        def __getattr__(self, alias):
+            project_id = self.dataset.project_id
```

Comment:
I'd vote for raising errors and correct metadata. We have to be able to make *some* reasonable assumptions about the files we want to process. Otherwise, our code becomes a rats nest of codified corrections that get out-of-date and don't work on the next batch of files.